RealEstateDataCleaning.py

```diff
@@ -8,113 +8,160 @@
         return '1 Story'
     else:
         return '2 Story'
 
 def extract_dummies(df, features, col):
-    '''Takes a list and makes a column for each phrase in that list
-    it places a 1 in that column if that phrase is present in the description else it places a 0 
+    """Takes a list and makes a column for each phrase in that list
+    it places a 1 in that column if that phrase is present in the description else it places a 0
 
     Args:
         df (pandas DataFrame): The DataFrame containing the descriptions
         features (list): The list of phrases that are being looked for in descriptions
-        col (str): name of column which contains the descriptions  
+        col (str): name of column which contains the descriptions
 
     Returns:
-        pandas DataFrame: without the column which contains descriptions plus 
-        additional columns one for each word in features 
-        
-    '''
-    
+        pandas DataFrame: without the column which contains descriptions plus
+        additional columns one for each word in features
+
+    """
+
     for feature in features:
         df[feature] = np.where((df[col].str.contains(feature)),1,0)
     df.drop([col], axis=1, inplace=True)
     return df
 
-def main():
-	df = pd.read_csv(input_file)
-	df.drop(df[df.SqFtTotal == 0].index, inplace=True)
-	df['CloseDate']  = pd.to_datetime(df['CloseDate'])
-	df = concatenate_address(df)
-	df = extract_dummies(df, Exterior_features, col='Exterior')
-	df['BasementDesc'] = df['BasementDesc'].apply(Basement_cleanup)
-	df = extract_dummies(df, BasementDesc_features, col='BasementDesc')
-	df = extract_dummies(df, ParkingDesc_features, col='ParkingDesc')
-	df['Stories'] = df['Stories'].apply(Stories_cleanup) 
-	df = extract_dummies(df, Stories_features, col='Stories')
-	df = make_YearBuilt_bins(df, size=10)
-	df['Year'] = pd.DatetimeIndex(df['CloseDate']).year
-	df['Age'] = df['Year'] - df['YearBuilt']
-	df['AssociationFee'] = np.where((df['AssociationFee']> 0),1,0)
-	df['PoolonProperty'] = np.where((df['PoolonProperty'] !=  'None'),1,0)
-	df['Bedrooms'] = df['UpperBedrooms'] + df['MainBedrooms']
-	df['FullBaths'] = df['UpperFullBaths'] + df['MainFullBaths']
-	df['HalfBaths'] = df['UpperHalfBaths'] + df['MainHalfBaths']
-	df.drop(['StreetNumber', 'StreetName', 'City', 'CountyOrParish', 'StateOrProvince', 
-			 'PostalCode','UpperBedrooms', 'MainBedrooms', 'UpperFullBaths', 'MainFullBaths', 
-			 'UpperHalfBaths', 'MainHalfBaths'], axis=1, inplace=True)
-	df.to_csv(output_file)
+def add_upper_and_main(data_frame, desired):
+    """
+    Look for the "Upper" and "Main" values for the desired column.  Sum these together and
+    create a new column named desired.  Finally, remove the "Upper" and "Main" columns
+
+    e.g. "Bedrooms" is the sum of the "UpperBedrooms" and "MainBedrooms" values
+
+    :param data_frame: callback frame object to manipulate
+    :param desired: string denoting the desired domain
+    :return: None
+    """
```

Comment:
So you asked about docstrings.  Docstrings are morally similar to JavaDoc.  They describe the function and define its API.  They are not, however, code, and they have a tendency to drift.  That is: it's easy to add or remove parameters from the function signature and forget to update the docstring.  Or you change something important about the function (such as, it modifies one of the values passed in) and you don't update the doc string.  The only thing worse than no docs are incorrect docs.

You said something about needing docstrings, though, so I am guessing they are mandated.  You'll see here that `:param` and `:return` stuff.  IntelliJ helpfully filled that in for me.

Your docstring should have a few things:

- a concise but accurate description of what the function does
- anything surprising or dangerous that might happen ("Warning: this will delete files on your local disk, use with caution")
- possibly an example
- a list of the parameters and their data types.  'callback' here is code for 'the parameter will be changed
- the return value, if any
- any exceptions that you would expect to be thrown